pkg/net/grpc/server/logInterceptors.go

```diff
@@ -0,0 +1,316 @@
+
+	reqData := logGrpcMessage{
+		StreamID: streamID,
+	}
+	reqData.Service, reqData.Method = parseServiceAndMethod(fullMethod)
+	sub := getSub(ctx)
+	if sub != "" {
+		reqData.JWT = &jwtMember{
+			Sub: sub,
+		}
+	}
+	if dumpBody {
+		reqData.Body = DecodeToJsonObject(req)
+	}
+	if !reqData.IsEmpty() {
+		logger = logger.With(log.RequestKey, reqData)
+	}
+
+	respData := logGrpcMessage{}
+	if code != nil {
+		respData.Code = code.String()
+	}
+	if err != nil {
+		respData.Error = err.Error()
+	}
+	if dumpBody {
+		respData.Body = DecodeToJsonObject(resp)
+	}
+	if !respData.IsEmpty() {
+		logger = logger.With(log.ResponseKey, respData)
+	}
+
+	return logger
+}
+
+func handleUnary(ctx context.Context, logger log.Logger, dumpBody bool, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	startTime := time.Now()
+	resp, err := handler(ctx, req)
+	code := status.Code(err)
+	level := DefaultCodeToLevel(code)
+	if !logger.Check(level) {
+		return resp, err
+	}
+	duration := time.Since(startTime)
+	logger = logUnary(ctx, logger, req, resp, &code, err, info.FullMethod, dumpBody, startTime, duration, "")
+	logger.GetLogFunc(level)("finished unary call with code " + code.String())
+	return resp, err
+}
+
+func NewLogUnaryServerInterceptor(logger log.Logger, dumpBody bool) grpc.UnaryServerInterceptor {
+	logger = logger.With(log.ProtocolKey, "GRPC")
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		return handleUnary(ctx, logger, dumpBody, req, info, handler)
+	}
+}
+
+// serverStream wraps around the embedded grpc.ServerStream, and intercepts the RecvMsg and
+// SendMsg method call.
+type logServerStream struct {
+	grpc.ServerStream
+	logger   log.Logger
+	dumpBody bool
+	streamID string
+	service  string
+	method   string
+	sub      string
+}
+
+func (w *logServerStream) SendMsg(m interface{}) error {
+	err := w.ServerStream.SendMsg(m)
+	code := status.Code(err)
+	level := DefaultCodeToLevel(code)
+	if errors.Is(err, io.EOF) {
+		level = log.DebugLevel
+	}
+	if !w.logger.Check(level) {
+		return err
+	}
+	logger := w.getLoggerForStreamMessage(m, err)
+	logger.Debug("sent a streaming message")
+	return err
+}
+
+func (w *logServerStream) getLoggerForStreamMessage(m interface{}, err error) log.Logger {
+	r := logGrpcMessage{
+		Service:  w.service,
+		Method:   w.method,
+		StreamID: w.streamID,
+	}
+	if err != nil {
+		r.Error = err.Error()
+	}
+	if w.sub != "" {
+		r.JWT = &jwtMember{
+			Sub: w.sub,
+		}
+	}
+	if w.dumpBody {
+		r.Body = DecodeToJsonObject(m)
+	}
+	logger := w.logger
+	logger = FillLoggerWithDeviceIDHrefCorrelationID(logger, m)
+	if !r.IsEmpty() {
+		logger = logger.With(log.MessageKey, r)
+	}
+	return logger
+}
+
+func (w *logServerStream) RecvMsg(m interface{}) error {
+	err := w.ServerStream.RecvMsg(m)
+	code := status.Code(err)
+	level := DefaultCodeToLevel(code)
+	if errors.Is(err, io.EOF) {
+		return err
```

Comment:
The code in `SendMsg` is very similar, expect there the code is like this:

```go
	if errors.Is(err, io.EOF) {
		level = log.DebugLevel
	}
````

Is the difference deliberate?

Reply:
Yes. When recv gets error EOF it means that client sends close message. In send if you get EOF this means that you was not able to send message to client so we want to log which message was not send.